In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import keras
from keras.applications.vgg19 import VGG19
from keras.models import Model
from keras.layers import Dense, Dropout, Flatten

import tensorflow as tf
import random
import os
from tqdm import tqdm
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
import cv2
# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

from subprocess import check_output
print(check_output(["ls", "../input"]).decode("utf8"))

# Any results you write to the current directory are saved as output.

Using TensorFlow backend.


labels.csv
sample_submission.csv
test
train



In [2]:
# parameters
dontFreezeLast = 2;

patience = 30;

loadWeights = False;
saveWeights = False;

tensorboard_dir = '../tb/VGG19v1/best_adam_opt';
gpuName = '/device:GPU:0'
tensorboardFlag = True;
workers = 10;
histogram_freq = 0;

modelPath = '../models/VGG19_opt/run_adam.h5';

batchSize = 64;
epochs = 100;
validation_size=0.3;

First we will read in the csv's so we can see some more information on the filenames and breeds

In [3]:
df_train = pd.read_csv('../input/labels.csv')
df_test = pd.read_csv('../input/sample_submission.csv')

print('Training images: ',df_train.shape[0])
print('Test images: ',df_test.shape[0])

# reduce dimensionality
#df_train = df_train.head(100)
#df_test = df_test.head(100)

Training images:  10222
Test images:  10357


In [4]:
df_train.head(10)

id               breed
0  000bec180eb18c7604dcecc8fe0dba07         boston_bull
1  001513dfcb2ffafc82cccf4d8bbaba97               dingo
2  001cdf01b096e06d78e9e5112d419397            pekinese
3  00214f311d5d2247d5dfe4fe24b2303d            bluetick
4  0021f9ceb3235effd7fcde7f7538ed62    golden_retriever
5  002211c81b498ef88e1b40b9abf84e1d  bedlington_terrier
6  00290d3e1fdd27226ba27a8ce248ce85  bedlington_terrier
7  002a283a315af96eaea0e28e7163b21b              borzoi
8  003df8b8a8b05244b1d920bb6cf451f9             basenji
9  0042188c895a2f14ef64a918ed9c7b64  scottish_deerhound

We can see that the breed needs to be one-hot encoded for the final submission, so we will now do this.

In [5]:
targets_series = pd.Series(df_train['breed'])
one_hot = pd.get_dummies(targets_series, sparse = True)

In [6]:
one_hot_labels = np.asarray(one_hot)

Next we will read in all of the images for test and train, using a for loop through the values of the csv files. I have also set an im_size variable which sets the size for the image to be re-sized to, 90x90 px, you should play with this number to see how it affects accuracy.

In [7]:
im_size = 90

In [8]:
x_train = []
y_train = []
x_test = []

In [9]:
i = 0 
for f, breed in tqdm(df_train.values):
    img = cv2.imread('../input/train/{}.jpg'.format(f))
    label = one_hot_labels[i]
    x_train.append(cv2.resize(img, (im_size, im_size)))
    y_train.append(label)
    i += 1

100%|██████████| 10222/10222 [00:49<00:00, 204.87it/s]


In [10]:
# for f in tqdm(df_test['id'].values):
#     img = cv2.imread('../input/test/{}.jpg'.format(f))
#     x_test.append(cv2.resize(img, (im_size, im_size)))

In [11]:
y_train_raw = np.array(y_train, np.uint8)
x_train_raw = np.array(x_train, np.float32) / 255.
x_test  = np.array(x_test, np.float32) / 255.

We check the shape of the outputs to make sure everyting went as expected.

In [12]:
print(x_train_raw.shape)
print(y_train_raw.shape)
print(x_test.shape)

(10222, 90, 90, 3)
(10222, 120)
(0,)


We can see above that there are 120 different breeds. We can put this in a num_class variable below that can then be used when creating the CNN model.

In [13]:
num_class = y_train_raw.shape[1]
print('Number of classes: ', num_class)

Number of classes:  120


It is important to create a validation set so that you can gauge the performance of your model on independent data, unseen to the model in training. We do this by splitting the current training set (x_train_raw) and the corresponding labels (y_train_raw) so that we set aside 30 % of the data at random and put these in validation sets (X_valid and Y_valid).

* This split needs to be improved so that it contains images from every class, with 120 separate classes some can not be represented and so the validation score is not informative.

In [14]:
X_train, X_valid, Y_train, Y_valid = train_test_split(x_train_raw, y_train_raw, test_size=validation_size, random_state=1)

Now we build the CNN architecture. Here we are using a pre-trained model VGG19 which has already been trained to identify many different dog breeds (as well as a lot of other objects from the imagenet dataset see here for more information: http://image-net.org/about-overview). Unfortunately it doesn't seem possible to downlod the weights from within this kernel so make sure you set the weights argument to 'imagenet' and not None, as it currently is below.

We then remove the final layer and instead replace it with a single dense layer with the number of nodes corresponding to the number of breed classes we have (120).

In [15]:
# Create the base pre-trained model
# Can't download weights in the kernel
with tf.device(gpuName):
    base_model = VGG19(#weights='imagenet',
        weights = 'imagenet', include_top=False, input_shape=(im_size, im_size, 3))

    # Add a new top layer
    x = base_model.output
    x = Flatten()(x)
    x = Dropout(0.7)(x)
    x = Dense(1024,activation='relu')(x)
    x = Dense(512,activation='relu')(x)
    predictions = Dense(num_class, activation='softmax')(x)

    # This is the model we will train
    model = Model(inputs=base_model.input, outputs=predictions)

    # First: train only the top layers (which were randomly initialized)
    for i in range(len(base_model.layers)-dontFreezeLast):
        base_model.layers[i].trainable = False

    if loadWeights:
        model.load_weights(checkPointPath)
    
    lr=0.00020389590556056983;
    beta_1=0.9453158868247398;
    beta_2=0.9925872692991417;
    decay=0.000821336141287018;
    adam = keras.optimizers.Adam(lr=lr,beta_1=beta_1,beta_2=beta_2,decay=decay)
    model.compile(loss='categorical_crossentropy', 
                  optimizer=adam, 
                  metrics=['accuracy'])

    callbacks_list = [];
    callbacks_list.append(keras.callbacks.EarlyStopping(
        monitor='val_acc',
        patience=patience,
        verbose=1));
    if saveWeights:
        callbacks_list.append(keras.callbacks.ModelCheckpoint(
            checkPointPath,
            monitor='val_acc',
            verbose=1,
            save_best_only=True,
            mode='max',
            save_weights_only=True))
    if tensorboardFlag:
        callbacks_list.append(keras.callbacks.TensorBoard(
                log_dir=tensorboard_dir,
                histogram_freq=histogram_freq,
                write_graph=False,
                write_images=False));
        print('Tensorboard activated in directory: ',tensorboard_dir)
    else:
        print('Tensorboard NOT activated')


    model.summary()

Tensorboard activated in directory:  ../tb/VGG19v1/best_adam_opt
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         (None, 90, 90, 3)         0         
_________________________________________________________________
block1_conv1 (Conv2D)        (None, 90, 90, 64)        1792      
_________________________________________________________________
block1_conv2 (Conv2D)        (None, 90, 90, 64)        36928     
_________________________________________________________________
block1_pool (MaxPooling2D)   (None, 45, 45, 64)        0         
_________________________________________________________________
block2_conv1 (Conv2D)        (None, 45, 45, 128)       73856     
_________________________________________________________________
block2_conv2 (Conv2D)        (None, 45, 45, 128)       147584    
_________________________________________________________________
block2_pool

In [16]:
def generator(X,Y,batch_size):
    batch_features = np.ndarray(shape=(batch_size,) + X.shape[1:],
                                dtype=X.dtype);
    batch_labels = np.ndarray(shape=(batch_size,) + Y.shape[1:],
                                dtype=Y.dtype)
    N = X.shape[0];
    
    while True:
        for i in range(batch_size):
            # choose random index in features
            index= np.random.choice(N,1)
            batch_features[i] = X[index]
            batch_labels[i] = Y[index]
        yield batch_features, batch_labels

In [17]:
model.fit(X_train, Y_train,
          epochs=epochs,
          batch_size = batchSize,
          validation_data=(X_valid, Y_valid),
          verbose=1,
          callbacks=callbacks_list)
model.save(modelPath);
# steps_per_epoch = round(X_train.shape[0]/batchSize)
# model.fit_generator(generator(X_train,Y_train,batchSize),
#                    steps_per_epoch=steps_per_epoch,
#                    epochs=epochs,
#                    verbose=1,
#                    callbacks=callbacks_list,
#                    validation_data=(X_valid,Y_valid),
#                    workers=workers,
#                    use_multiprocessing=True)

Train on 7155 samples, validate on 3067 samples
Epoch 1/100
7155/7155 [==============================] - 40s 6ms/step - loss: 4.8412 - acc: 0.0136 - val_loss: 4.7413 - val_acc: 0.0245
Epoch 2/100
7155/7155 [==============================] - 28s 4ms/step - loss: 4.6953 - acc: 0.0278 - val_loss: 4.6086 - val_acc: 0.0496
Epoch 3/100
7155/7155 [==============================] - 28s 4ms/step - loss: 4.4687 - acc: 0.0577 - val_loss: 4.3485 - val_acc: 0.0668
Epoch 4/100
7155/7155 [==============================] - 28s 4ms/step - loss: 4.1628 - acc: 0.0939 - val_loss: 4.0944 - val_acc: 0.0991
Epoch 5/100
7155/7155 [==============================] - 28s 4ms/step - loss: 3.8648 - acc: 0.1303 - val_loss: 3.9027 - val_acc: 0.1171
Epoch 6/100
7155/7155 [==============================] - 28s 4ms/step - loss: 3.6330 - acc: 0.1637 - val_loss: 3.7993 - val_acc: 0.1311
Epoch 7/100
7155/7155 [==============================] - 28s 4ms/step - loss: 3.4313 - acc: 0.1985 - val_loss: 3.7201 - val_acc: 0.1441


7155/7155 [==============================] - 28s 4ms/step - loss: 0.3890 - acc: 0.9071 - val_loss: 4.1907 - val_acc: 0.1947
Epoch 61/100
7155/7155 [==============================] - 28s 4ms/step - loss: 0.3650 - acc: 0.9154 - val_loss: 4.2267 - val_acc: 0.1963
Epoch 62/100
7155/7155 [==============================] - 28s 4ms/step - loss: 0.3711 - acc: 0.9121 - val_loss: 4.2199 - val_acc: 0.1956
Epoch 63/100
7155/7155 [==============================] - 28s 4ms/step - loss: 0.3482 - acc: 0.9196 - val_loss: 4.2365 - val_acc: 0.1982
Epoch 64/100
7155/7155 [==============================] - 28s 4ms/step - loss: 0.3360 - acc: 0.9245 - val_loss: 4.2573 - val_acc: 0.2005
Epoch 65/100
7155/7155 [==============================] - 28s 4ms/step - loss: 0.3341 - acc: 0.9219 - val_loss: 4.2425 - val_acc: 0.1989
Epoch 66/100
7155/7155 [==============================] - 28s 4ms/step - loss: 0.3334 - acc: 0.9251 - val_loss: 4.2501 - val_acc: 0.1943
Epoch 67/100
7155/7155 [==============================

Remember, accuracy is low here because we are not taking advantage of the pre-trained weights as they cannot be downloaded in the kernel. This means we are training the wights from scratch and I we have only run 1 epoch due to the hardware constraints in the kernel.

Next we will make our predictions.

In [18]:
# preds = model.predict(x_test, verbose=1)

In [19]:
# sub = pd.DataFrame(preds)
# # Set column names to those generated by the one-hot encoding earlier
# col_names = one_hot.columns.values
# sub.columns = col_names
# # Insert the column id from the sample_submission at the start of the data frame
# sub.insert(0, 'id', df_test['id'])
# sub.head(10)